### ISSUE TITLE
Provide plain SQL migrations for projects that do not use Ecto migrations runner

### ISSUE THREAD

**Issue by mkaput** [sorentwo/oban#653]
### Is your feature request related to a problem? Please describe.

My team is building a product which is architected as a set of services serving write ops for a central API Gateway which handles DB reads by itself. Each service connects to the **same** Postgres database, but lives in its own Postgres schema. Some services are written in Elixir, some are not.

![image](https://user-images.githubusercontent.com/3450050/155966120-1906ddf8-c231-4ab3-b092-5e611ba6473a.png)

The API Gateway works for us as the owner of our database schema, and we want to keep all database migrations there in order to avoid smearing logic all over our services codebases. Additionally, this theoretically should allow us to deny schema manipulation access to our services (we create a separate Postgres user for each service).

We are using Oban in some of these services. Oban jobs never escape from responsibility arena of each service, basically each service has its own independent Oban installation.

We are currently running Oban migrations as supported by the library. In order for that, we had to:
1. Grant each affected application schema manipulation permissions.
2. Make each application deployment phase run migrations only to get Oban up and running. Any other migration is managed by our central migration system.
3. Make each developer be very careful about configuring Ecto to use specified `prefix`. Surprisingly, we have learned that this is easy to overlook 😅 

### Describe the Solution You'd Like

This isn't tragic, but we find the fact that Oban requires us to migrate database using Ecto Migrations painful enough that we **consider** dropping these migrations and vendor Oban migrations ourselves in our central migrations repository. This means we will be manually translating Oban migrations to plain SQL. Yes, that's extra work, but I see that Oban isn't publishing new migrations frequently.

We would like to hear from you, what do you think about keeping plain SQL migrations in Oban repository for users to copy into their migration systems, and supporting this alternative path of installing Oban in database. Obviously, this would increase maintenance burden in Oban. The alternative approach could be documented in the [Installation guide](https://hexdocs.pm/oban/installation.html).

### Describe Alternatives You've Considered

1. Oban can do nothing here. We are fine with implementing this solution privately.
2. Oban may provide a Mix task (`mix oban.migrations.dump [version]`) which would automatically generate plain SQL from Elixir code. Ideally it would be awesome if Ecto provided such functionality built-in, but because Ecto allows arbitrary Elixir code in migrations, a general upstream solution is not possible IMHO.

### Caveats

1. This is a solution for a very specific scenario. I think it may be even exotic for general audience.
1. This approach requires more manual work. Documentation should make users aware of that and explicitly state that the benefits shine only in such scenarios.
1. Current migrations are parameterized by schema name (`prefix`). I think we could use `public.` prefix in all SQL statements and recommend to users running `sed` in order to put custom schema.
1. Current migrations system has a custom versioning approach, so it can automatically run many Oban migrations from a single Ecto migration step (for example, one Ecto migration may run `v4..v11` migration set internally). I think we should simply document that end users need to apply this logic in their systems manualy.
1. Implementing this approach would ban custom Elixir logic in future migrations.

**Comment by milmazz** [comment]
@mkaput As you already pointed out, this seems like such a specific scenario and it will add some burden to the maintainers, and also, even providing a raw SQL it will not be enough in many circumstances, as you also mentioned, in some cases, you will need to adapt the schema name or prefix.

So, have you thought about creating a temporal Ecto migration with your changes and then running something like `mix ecto.migrate --log-migrations-sql`?, the previous command will produce a log including the SQL generated by the migration commands. Or, you can create a local mix task based on `mix ecto.migrate  --log-migrations-sql` to produce what you need.

```console
$ mix ecto.migrate --log-migrations-sql

23:13:30.274 [info]  == Running 20220305051110 MyApp.Repo.Migrations.AddObanJobsTable.up/0 forward

23:13:30.294 [info]  execute "DO $$\nBEGIN\nIF NOT EXISTS (SELECT 1 FROM pg_type\n               WHERE typname = 'oban_job_state'\n                 AND typnamespace = 'public'::regnamespace::oid) THEN\n    CREATE TYPE public.oban_job_state AS ENUM (\n      'available',\n      'scheduled',\n      'executing',\n      'retryable',\n      'completed',\n      'discarded'\n    );\n  END IF;\nEND$$;\n"

23:13:30.439 [debug] QUERY OK db=138.6ms
DO $$
BEGIN
IF NOT EXISTS (SELECT 1 FROM pg_type
               WHERE typname = 'oban_job_state'
                 AND typnamespace = 'public'::regnamespace::oid) THEN
    CREATE TYPE public.oban_job_state AS ENUM (
      'available',
      'scheduled',
      'executing',
      'retryable',
      'completed',
      'discarded'
    );
  END IF;
END$$;
 []
....
```

Another approach, and it should produce what you want in a clean way, could be a sequence of `mix ecto.migrate` and `mix ecto.dump`, for example:

1. Before including the Oban migration, you run `mix ecto.migrate` to be up to date, and then you run `mix ecto.dump`, that Mix task will produce a dump file in `priv/repo/structure.sql`, backup this file.
2. Include the Oban migrations as recommended in the docs
3. Do prefix modifications if those apply in your case
4. Run again `mix ecto.migrate`
5. Run again `mix ecto.dump`
6. Diff between the generated `priv/repo/structure.sql` and your backup. That diff is the output you want.

HTH

**Comment by mkaput** [comment]
@milmazz I have thought about the `log` approach, but found it to be much messier than translating Elixir code to SQL by hand.

I have manually translated all migrations to SQL queries templated with EEx, so with a simple call to `EEx.eval_file` we can get a proper SQL statement set with all prefixes properly provided. On top of this, we have a little script which takes these migrations, concatenates them, writes `COMMENT` statement as [Oban's migrations module does](https://github.com/sorentwo/oban/blob/43f77510a3827a9f38fe67b3965b0fe62ee19c20/lib/oban/migrations.ex#L193-L197) and writes it to `up.sql` & `down.sql` files in our migrations too. We will deploy the first instances of these migrations this week and observe how it plays out.

**Comment by mkaput** [comment]
We have open sourced our scripts: https://github.com/btwb/oban_migrations_sql 🙂

**Comment by mkaput** [comment]
Hmm, I may not have spare bandwidth to merge these into Oban itself this week/in near future, so if somebody likes to do this, feel free 😄

**Comment by sorentwo** [comment]
@mkaput No rush. If you get to it in the next week or so, that's wonderful. Otherwise, I'll tackle it eventually 🙂

**Comment by azizk** [comment]
Maybe I don't completely understand the issue but I'm curious to learn and wanted to comment. I just don't quite see why it's worth it to convert all the migrations to SQL rather than come up with a more manageable solution. If I had this sort of architecture I would try to run Oban's migrations on the API gateway as well. How? Maybe by creating a simple Elixir app that always depends on latest Oban, which can execute the versioned migrations via a basic API endpoint (`/migrate?v=8&prefix=service2`). So then each service deployment script would call this endpoint with the right parameters. Problem solved? Maybe. You can only find out in practice. :slightly_smiling_face:

Just a thought. :wink:

**Comment by mkaput** [comment]
> Maybe I don't completely understand the issue but I'm curious to learn and wanted to comment. I just don't quite see why it's worth it to convert all the migrations to SQL rather than come up with a more manageable solution. If I had this sort of architecture I would try to run Oban's migrations on the API gateway as well. How? Maybe by creating a simple Elixir app that always depends on latest Oban, which can execute the versioned migrations via a basic API endpoint (`/migrate?v=8&prefix=service2`). So then each service deployment script would call this endpoint with the right parameters. Problem solved? Maybe. You can only find out in practice. :slightly_smiling_face:
> 
> 
> 
> Just a thought. :wink:

Seemingly smart idea but unfortunately I see two issues here:

1. You do not have migrations continuity encoded as linear set of changes here, because you run part of migrations _out-of-band_. How would you handle rolling back migrations here? How can external reader (say, new developer without access to the database) learn where Oban tables came from? You can answer these questions but these answers won't _smell good_.
2. Real life case -- our gateway is hosted on third party SaaS and we do not have shell access to the migration runner nor an ability to shell out.

Also IMHO, translating to SQL is just stupidly simpler as in KISS principle. And also there are fewer moving parts.

**Comment by sorentwo** [comment]
> translating to SQL is just stupidly simpler as in KISS principle. And also there are fewer moving parts.

This is the primary reason that I'm in favor of this change. It makes it easy to inspect what the migrations are doing, and it makes the files portable—meaning they can be used from any migration tool. That's more useful than you might expect for people who are transitioning between languages/migrators.

**Comment by azizk** [comment]
Regarding your points, @mkaput:
1. Not difficult I think. Devs should be safe. Just put the Oban version in a source/config file and reference it when sending the API request to the micro service `:oban_migrator` (as I'd call it). You could even do the API call in each `application.ex` on start-up. It should up- and down-migrate without issues and do nothing if the version didn't change (`Oban.Migrations` will noop). Of course, be careful and prevent retries on error.
2. I don't know. In that case you'll have to bite the bullet and live with the limitations. I'm not sure, but my solution may still work because it's simply a little service that can run the migrations dynamically. So it shouldn't matter if it's a SaaS?

I think my idea is pragmatic. It's not complicated and should also qualify as KISS afaics. However situations vary and so it depends.

Using pure SQL may be simple, but how disruptive is it actually going to be for all the users out there wanting to upgrade to the next version? Are there no cons? Perhaps it's better to write a guide for such special application architectures?

**Comment by mkaput** [comment]
@sorentwo I definitely won't have spare time to tackle this for next two/three weeks so just in case, please don't wait for me 😞

**Comment by mkaput** [comment]
@azizk I think we may misunderstand each other in one area. What me and @sorentwo want to achieve is to make **absolutely no** breaking changes for existing users with this feature. Existing Ecto migrations will work just like they used to do 🙂 

Ecto migrations interface is provided by [`Oban.Migrations`](https://github.com/sorentwo/oban/blob/ace5db1f384a780c9510ff30d7c8472a56e67695/lib/oban/migrations.ex) module. This module is expected to be running with Ecto migration context, so it uses Ecto's DSL, which includes the [`execute` function](https://hexdocs.pm/ecto_sql/Ecto.Migration.html#execute/1) (for example used [here](https://github.com/sorentwo/oban/blob/main/lib/oban/migrations.ex#L196)). Under the hood, these modules perform _Oban migration squashing_, so you (as end user) do not have to create separate Ecto migrations for each Oban's migration. Oban's migrations are currently plain Elixir modules (located [here](https://github.com/sorentwo/oban/tree/main/lib/oban/migrations)) which call Ecto's migration DSL.

What we will do is we will replace these modules with EEx templated raw SQL files, which will then be passed to `execute/2` call 🙂 

As a new feature, we will also provide a Mix Task that will allow you to dump these SQL templates as raw SQL queries, so you can feed them into other migration systems.

**Comment by sorentwo** [comment]
> I definitely won't have spare time to tackle this for next two/three weeks so just in case, please don't wait for me 😞

@mkaput No worries. There are some other efforts that I need to finish up before I can get to this.

**Comment by halostatue** [comment]
Hi, @sorentwo. I don’t have @mkaput’s specific scenario, but I _do_ have the problem presented. Specifically, for our main application, we use [`sqitch`](http://sqitch.org) as our database migration manager. There’s a _lot_ of reasons for this, but it’s something of an irreversible decision at this point. (I would not reverse it, by the way; as much as I like Ecto and its migrations are really good, the forced safety benefits of separating database migrations from code are _way_ too strong…and sqitch is a great fine migration manager.)

We have finally been able to upgrade our applications to a modern Elixir to be able to use Oban…and we need SQL and/or pgPL/SQL to be execute our migrations. My colleague @ohnit is going to be working on this, so if this is something that you want to consider integrating as suggested in this thread, I think we can work with you and @mkaput on that.

**Comment by mkaput** [comment]
@halostatue @ohnit never heard of `squitch`, but this sounds cool! From brief reading of their Postgres tutorial, it looks like they are just using plain pgSQL files, but directory structure is a bit different + I noticed the `-- requires: appschema` part. Are these the only differences?

If so, then short-term you should easily be able to use my existing code, and readjust conventions by hand. In the long term, I will be happy to quickly review and merge PRs to add support for new Sqitch formatter in my project, until it will be merged into Oban itself.

**Comment by halostatue** [comment]
The `-- requires:` bit is mostly documentation as the relationships between the various pieces are documented in the `sqitch.plan` file.

The main difference is that you create three files when generating for Sqitch:

1. Migration
2. Validation (a query that proves that your changes are present)
3. Rollback

The validation is _particularly_ valuable when testing that columns have been added correctly in a migration. (It’s a little harder to prove that columns have been correctly *removed*).

You also have the option of using [`pgTAP`](https://pgtap.org) to make _unit tests_ for the migrations, which is beneficial in several ways. I have had team members ask the purpose of validating what is essentially static code, but as we have evolved the tests, it turns out that we have caught some schema errors (fortunately, before going to production). It’s simplified for us because we take advantage of the templating that is available to generate most of the test code, but while these get generated for each migration, we keep them for each _table_ or function / procedure added where we can.

Obviously, for us, we would not need to generate pgTAP test files, but we will need to create (or have created) validation scripts, which basically boil down to "does the query being executed here throw an error". If they do, then the migration fails.

As an example, we have the following files to create the `users` table in our primary application.

<details><summary><strong>▶ deploy/users.sql</strong></summary>

```sql
-- Deploy platform-schema:users to pg
-- requires: organizations

BEGIN;

CREATE TABLE IF NOT EXISTS users (
  id uuid PRIMARY KEY,
  organization_id uuid NOT NULL,
  creator_id uuid,
  status user_status NOT NULL,
  inserted_at timestamp without time zone NOT NULL DEFAULT now(),
  updated_at timestamp without time zone NOT NULL DEFAULT now(),

  CONSTRAINT users_organization_id_fkey
    FOREIGN KEY (organization_id)
    REFERENCES organizations(id)
    ON DELETE CASCADE DEFERRABLE INITIALLY DEFERRED,
  CONSTRAINT users_creator_id_fkey
    FOREIGN KEY (creator_id)
    REFERENCES users(id)
    ON DELETE CASCADE DEFERRABLE INITIALLY DEFERRED
);

COMMIT;
```

</details>

<details><summary><strong>▶ verify/users.sql</strong></summary>

```sql
-- Verify platform-schema:users on pg

BEGIN;

SELECT id, organization_id, creator_id, status, inserted_at, updated_at
  FROM users
 WHERE FALSE;

ROLLBACK;
```

</details>

<details><summary><strong>▶ test/users.sql</strong></summary>

```sql
-- Test kinetic-platform-schema:users on pg

SET client_min_messages TO warning;
RESET client_min_messages;
SET search_path TO public;

BEGIN;
SELECT plan(38);

SELECT has_table('users');
SELECT has_pk('users');

SELECT columns_are(
  'users',
  ARRAY[
    $$id$$,
    $$organization_id$$,
    $$creator_id$$,
    $$status$$,
    $$inserted_at$$,
    $$updated_at$$,
    $$profile_credential_id$$
  ]
);

\set col $$users$$, $$id$$

SELECT has_column(:col);
SELECT col_type_is(:col, 'uuid');
SELECT col_is_pk(:col);

\set col $$users$$, $$organization_id$$

SELECT has_column(:col);
SELECT col_type_is(:col, 'uuid');
SELECT col_hasnt_default(:col);
SELECT col_is_fk(:col);
SELECT col_not_null(:col);
SELECT fk_ok(:col, 'organizations', 'id');

\set col $$users$$, $$creator_id$$

SELECT has_column(:col);
SELECT col_type_is(:col, 'uuid');
SELECT col_hasnt_default(:col);
SELECT col_is_fk(:col);
SELECT col_is_null(:col);
SELECT fk_ok(:col, 'users', 'id');

\set col $$users$$, $$status$$

SELECT has_column(:col);
SELECT col_type_is(:col, 'user_status');
SELECT col_not_null(:col);
SELECT col_hasnt_default(:col);

\set col $$users$$, $$inserted_at$$

SELECT has_column(:col);
SELECT col_type_is(:col, 'timestamp without time zone');
SELECT col_not_null(:col);
SELECT col_has_default(:col);
SELECT col_default_is(:col, $$now()$$);

\set col $$users$$, $$updated_at$$

SELECT has_column(:col);
SELECT col_type_is(:col, 'timestamp without time zone');
SELECT col_not_null(:col);
SELECT col_has_default(:col);
SELECT col_default_is(:col, $$now()$$);

\set col $$users$$, $$profile_credential_id$$

SELECT has_column(:col);
SELECT col_type_is(:col, 'uuid');
SELECT col_hasnt_default(:col);
SELECT col_is_fk(:col);
SELECT col_is_null(:col);
SELECT fk_ok(:col, 'user_credentials', 'id');


SELECT finish();
ROLLBACK;
```

</details>

<details><summary><strong>▶ revert/users.sql</strong></summary>

```sql
--- revert/users.sql
-- Revert kinetic-platform-schema:users from pg

BEGIN;

DROP TABLE users;

COMMIT;
```

</details>

**Comment by mkaput** [comment]
Oban's migrations do not have verification queries, so that would be something new here. Frankly, personally, I would be hesitant to incorporate this concept to current setup as that would increase maintenance burden. I wonder if it wouldn't be enough to just leave verification query empty?

**Comment by sorentwo** [comment]
@halostatue Thank you for the additional use case validation and context. The additional validation work is very cool, but not something I'd like to include in Oban itself. The migrations already have a degree of testing through the migration tests and other parts of the test suite.

This is still something I want to integrate, but there have been a few distractions (mostly prepping for our virtual training in ElixirConfEU). A PR is definitely welcome!

**Comment by sorentwo** [comment]
After working on this for a while, I've encountered an issue with grouping/executing SQL statements. Namely, combining multiple commands within a migration causes this error:

```
ERROR 42601 (syntax_error) cannot insert multiple commands into a prepared statement
```

The workaround is to wrap every SQL command in a `DO/END` statement, which I find cumbersome and noisy. The result is SQL files with an unusual structure, and a lot of `<%= @prefix %>` interpolation, which I don't think is helpful for external projects.

I'm leaning toward linking to the excellent `oban_migrations_sql` package for people that want to use an external runner and documenting the necessary commands to get a truly plain SQL dump.

**Comment by conradfr** [comment]
What a shame this was not taken seriously and implemented.

And the open source migration has not been updated (it goes to v11 for reference)

**Comment by mkaput** [comment]
@conradfr PRs are welcome :)

**Comment by conradfr** [comment]
Well for anyone arriving here, here is the sql I used, going up to v13

```
-- v01

CREATE SCHEMA IF NOT EXISTS public;

DO
$$
    BEGIN
        IF NOT EXISTS(SELECT 1
                      FROM pg_type
                      WHERE typname = 'oban_job_state'
                        AND typnamespace = 'public'::regnamespace::oid) THEN
            CREATE TYPE public.oban_job_state AS ENUM (
                'available',
                'scheduled',
                'executing',
                'retryable',
                'completed',
                'discarded'
                );
        END IF;
    END
$$;

CREATE TABLE IF NOT EXISTS public.oban_jobs
(
    id           BIGSERIAL PRIMARY KEY,
    state        public.oban_job_state     NOT NULL DEFAULT 'available',
    queue        TEXT                        NOT NULL DEFAULT 'default',
    worker       TEXT                        NOT NULL,
    args         JSONB                       NOT NULL,
    errors       JSONB[]                     NOT NULL DEFAULT ARRAY []::JSONB[],
    attempt      INTEGER                     NOT NULL DEFAULT 0,
    max_attempts INTEGER                     NOT NULL DEFAULT 20,
    inserted_at  TIMESTAMP WITHOUT TIME ZONE NOT NULL DEFAULT timezone('UTC', NOW()),
    scheduled_at TIMESTAMP WITHOUT TIME ZONE NOT NULL DEFAULT timezone('UTC', NOW()),
    attempted_at TIMESTAMP WITHOUT TIME ZONE,
    completed_at TIMESTAMP WITHOUT TIME ZONE
);

CREATE INDEX IF NOT EXISTS oban_jobs_queue_index ON public.oban_jobs (queue);
CREATE INDEX IF NOT EXISTS oban_jobs_state_index ON public.oban_jobs (state);
CREATE INDEX IF NOT EXISTS oban_jobs_scheduled_at_index ON public.oban_jobs (scheduled_at);

CREATE OR REPLACE FUNCTION public.oban_jobs_notify() RETURNS trigger AS
$$
DECLARE
    channel text;
    notice  json;
BEGIN
    IF (TG_OP = 'INSERT') THEN
        channel = 'public.oban_insert';
        notice = JSON_BUILD_OBJECT('queue', NEW.queue, 'state', NEW.state);
        -- No point triggering for a job that isn't scheduled to run now
        IF NEW.scheduled_at IS NOT NULL AND NEW.scheduled_at > NOW() AT TIME ZONE 'utc' THEN
            RETURN NULL;
        END IF;
    ELSE
        channel = 'public.oban_update';
        notice = JSON_BUILD_OBJECT('queue', NEW.queue, 'new_state', NEW.state, 'old_state',
                                   OLD.state);
    END IF;
    PERFORM pg_notify(channel, notice::text);
    RETURN NULL;
END;
$$ LANGUAGE plpgsql;

DROP TRIGGER IF EXISTS oban_notify ON public.oban_jobs;

CREATE TRIGGER oban_notify
    AFTER INSERT OR UPDATE OF state
    ON public.oban_jobs
    FOR EACH ROW
EXECUTE PROCEDURE public.oban_jobs_notify();

-- v02
DROP INDEX IF EXISTS public.oban_jobs_scheduled_at_index;

CREATE INDEX IF NOT EXISTS oban_jobs_scheduled_at_index ON public.oban_jobs (scheduled_at)
    WHERE state IN ('available'::public.oban_job_state, 'scheduled'::public.oban_job_state);

ALTER TABLE public.oban_jobs
    ADD CONSTRAINT queue_length CHECK (((CHAR_LENGTH(queue) > 0) AND (CHAR_LENGTH(queue) < 128))),
    ADD CONSTRAINT worker_length CHECK (((CHAR_LENGTH(worker) > 0) AND (CHAR_LENGTH(worker) < 128)));

CREATE OR REPLACE FUNCTION public.oban_wrap_id(value bigint) RETURNS int AS
$$
BEGIN
    RETURN (CASE WHEN value > 2147483647 THEN MOD(value, 2147483647) ELSE value END)::int;
END;
$$ LANGUAGE plpgsql IMMUTABLE;

-- v03
ALTER TABLE public.oban_jobs
    ADD COLUMN IF NOT EXISTS attempted_by text[];

-- v04
DROP FUNCTION IF EXISTS public.oban_wrap_id(VALUE bigint);

-- v05
DROP INDEX IF EXISTS public.oban_jobs_queue_index;
DROP INDEX IF EXISTS public.oban_jobs_state_index;
DROP INDEX IF EXISTS public.oban_jobs_scheduled_at_index;

CREATE INDEX IF NOT EXISTS oban_jobs_queue_state_scheduled_at_id_index
    ON public.oban_jobs (queue, state, scheduled_at, id);

-- v06
-- This used to modify oban_beats, which aren't included anymore

-- v07
CREATE INDEX IF NOT EXISTS oban_jobs_attempted_at_id_index
    ON public.oban_jobs (attempted_at DESC, id)
    WHERE state IN ('completed', 'discarded');

-- v08
ALTER TABLE public.oban_jobs
    ADD COLUMN IF NOT EXISTS discarded_at timestamp WITHOUT TIME ZONE,
    ADD COLUMN IF NOT EXISTS priority     integer DEFAULT 0,
    ADD COLUMN IF NOT EXISTS tags         character varying(255)[] DEFAULT ARRAY[]::character varying[];

DROP INDEX IF EXISTS public.oban_jobs_queue_state_scheduled_at_id_index;

CREATE INDEX IF NOT EXISTS oban_jobs_state_queue_priority_scheduled_at_id_index
    ON public.oban_jobs (state, queue, priority, scheduled_at, id);

CREATE OR REPLACE FUNCTION public.oban_jobs_notify() RETURNS trigger AS
$$
DECLARE
    channel text;
    notice  json;
BEGIN
    IF NEW.state = 'available' THEN
        channel = 'public.oban_insert';
        notice = JSON_BUILD_OBJECT('queue', NEW.queue);
        PERFORM pg_notify(channel, notice::text);
    END IF;
    RETURN NULL;
END;
$$ LANGUAGE plpgsql;

DROP TRIGGER IF EXISTS oban_notify ON public.oban_jobs;

CREATE TRIGGER oban_notify
    AFTER INSERT
    ON public.oban_jobs
    FOR EACH ROW
EXECUTE PROCEDURE public.oban_jobs_notify();

-- v09
ALTER TABLE public.oban_jobs
    ADD COLUMN IF NOT EXISTS meta         jsonb DEFAULT '{}'::jsonb,
    ADD COLUMN IF NOT EXISTS cancelled_at timestamp WITHOUT TIME ZONE;

DO
$$
    DECLARE
        version int;
        already bool;
    BEGIN
        SELECT CURRENT_SETTING('server_version_num')::int INTO version;
        SELECT '{cancelled}' <@ ENUM_RANGE(NULL::public.oban_job_state)::text[] INTO already;
        IF already THEN
            RETURN;
        ELSIF version >= 120000 THEN
            ALTER TYPE public.oban_job_state ADD VALUE IF NOT EXISTS 'cancelled';
        ELSE
            ALTER TYPE public.oban_job_state RENAME TO old_oban_job_state;
            CREATE TYPE public.oban_job_state AS ENUM (
                'available',
                'scheduled',
                'executing',
                'retryable',
                'completed',
                'discarded',
                'cancelled'
                );
            ALTER TABLE public.oban_jobs
                RENAME COLUMN state TO _state;
            ALTER TABLE public.oban_jobs
                ADD state public.oban_job_state NOT NULL DEFAULT 'available';
            UPDATE public.oban_jobs SET state = _state::text::public.oban_job_state;
            ALTER TABLE public.oban_jobs
                DROP COLUMN _state;
            DROP TYPE public.old_oban_job_state;
        END IF;
    END
$$;

CREATE INDEX IF NOT EXISTS oban_jobs_state_queue_priority_scheduled_at_id_index
    ON public.oban_jobs (state, queue, priority, scheduled_at, id);

-- v10
ALTER TABLE public.oban_jobs
    -- These didn't have defaults out of consideration for older PG versions
    ALTER COLUMN args SET DEFAULT '{}'::jsonb,
    ALTER COLUMN priority SET NOT NULL,

    -- These could happen from an insert_all call with invalid data
    ADD CONSTRAINT attempt_range CHECK (((attempt >= 0) AND (attempt <= max_attempts))),
    ADD CONSTRAINT positive_max_attempts CHECK ((max_attempts > 0)),
    ADD CONSTRAINT priority_range CHECK (((priority >= 0) AND (priority <= 3)));

DROP INDEX IF EXISTS public.oban_jobs_args_vector;
DROP INDEX IF EXISTS public.oban_jobs_worker_gist;
DROP INDEX IF EXISTS public.oban_jobs_attempted_at_id_index;

CREATE INDEX IF NOT EXISTS oban_jobs_args_index ON public.oban_jobs USING gin (args);
CREATE INDEX IF NOT EXISTS oban_jobs_meta_index ON public.oban_jobs USING gin (meta);

-- v11
CREATE UNLOGGED TABLE IF NOT EXISTS public.oban_peers
(
    name       text                        NOT NULL PRIMARY KEY,
    node       text                        NOT NULL,
    started_at timestamp WITHOUT TIME ZONE NOT NULL,
    expires_at timestamp WITHOUT TIME ZONE NOT NULL
);

ALTER TABLE "public".oban_peers SET UNLOGGED;

-- v12
ALTER TABLE "public"."oban_jobs" DROP CONSTRAINT "priority_range";
ALTER TABLE "public"."oban_jobs" ADD CONSTRAINT "non_negative_priority" CHECK (priority >= 0) NOT VALID;

DROP TRIGGER IF EXISTS oban_notify ON "public".oban_jobs;
DROP FUNCTION IF EXISTS "public".oban_jobs_notify();

-- v13
CREATE INDEX IF NOT EXISTS "oban_jobs_state_cancelled_at_index" ON "public"."oban_jobs" ("state", "cancelled_at");
CREATE INDEX IF NOT EXISTS "oban_jobs_state_discarded_at_index" ON "public"."oban_jobs" ("state", "discarded_at");

COMMENT ON TABLE "public".oban_jobs IS '13'
```
